docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,340 @@
+   * [Proposal](#proposal)
+      * [IPAM API Contract](#ipam-api-contract)
+      * [Pools &amp; IPAM Providers](#pools--ipam-providers)
+      * [Consumption](#consumption)
+      * [User Stories](#user-stories)
+         * [Story 1](#story-1)
+         * [Story 2](#story-2)
+         * [Story 3](#story-3)
+         * [Story 4](#story-4)
+      * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Security Model](#security-model)
+      * [Risks and Mitigations](#risks-and-mitigations)
+   * [Alternatives](#alternatives)
+   * [Upgrade Strategy](#upgrade-strategy)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+### IPAM Provider
+
+A controller that watches IPAddressClaims and fulfils them with IPAddresses that it allocates from an IP Pool. It comes with it's own IP Pool custom resource definition to provide parameters for allocating addresses. Providers can work in-cluster relying only on custom resources, or integrate with external systems like Netbox or Infoblox.
+
+### IPAddressClaim
+
+An IPAddressClaim is a custom resource that can be used by infrastructure providers to allocate IP addresses from IP Pools that are managed by IPAM providers.
+
+### IPAddress
+
+An IPAddress is a custom resource that gets created by an IPAM provider to fulfil an IPAddressClaim, which then gets consumed by the infrastructure provider that created the claim.
+
+### IP Pool
+
+An IP Pool is a custom resource that is provided by a IPAM provider, which holds configuration for a single pool of IP addresses that can be allocated from using an IPAddressClaim.
+
+## Summary
+
+This proposal adds an API contract for pluggable IP Address Management to CAPI that any infrastructure provider can choose to implement. It allows them to dynamically allocate and release IP Addresses from IP Address Pools following a similar concept as Kubernetes' PersistentVolumes. Pools are managed by IPAM Providers, which allow to integrate with external systems like Netbox or Infoblox.
+
+## Motivation
+
+IP address management for machines is currently left to the infrastructure providers for CAPI. Most on-premise providers (e.g. CAPV) allow the use of either DHCP or statically configured IPs. Since Machines are created from templates, static allocation would require a single template for each machine, which prevents dynamic scaling of nodes without custom controllers that create new templates.
+
+While DHCP is a viable solution for dynamic IP assignment which also enables scaling, it can cause problems in conjunction with CAPI. Especially in smaller networks rolling cluster upgrades can exhaust the network in use. When multiple machines are replaced in quick succession, each of them will get a new DHCP lease. Unless the lease time is very short, at least twice as many IPs as the maximum number of nodes has to be allocated to a cluster.
```

Comment:
can we elaborate a bit how this proposal alleviate this problem?

Reply:
I've added a short paragraph in the Proposal section.